README
NAME
    Test::Reporter - sends test results to cpan-testers@perl.org

SYNOPSIS
      use Test::Reporter;

      my $reporter = Test::Reporter->new();

      $reporter->grade('pass');
      $reporter->distribution('Mail-Freshmeat-1.20');
      $reporter->send() || die $reporter->errstr();

      # or

      my $reporter = Test::Reporter->new();

      $reporter->grade('fail');
      $reporter->distribution('Mail-Freshmeat-1.20');
      $reporter->comments('output of a failed make test goes here...');
      $reporter->edit_comments(); # if you want to edit comments in an editor
      $reporter->send() || die $reporter->errstr();

      # or

      my $reporter = Test::Reporter->new(
          grade => 'fail',
          distribution => 'Mail-Freshmeat-1.20',
          from => 'whoever@wherever.net (Whoever Wherever)',
          comments => 'output of a failed make test goes here...',
          via => 'CPANPLUS X.Y.Z',
      );
      $reporter->send() || die $reporter->errstr();

DESCRIPTION
    Test::Reporter reports the test results of any given distribution to the
    CPAN Testers. Test::Reporter has wide support for various perl5's and
    platforms. For further information visit the below links:

    *   <http://www.cpantesters.org/>

        CPAN Testers reports (new site)

    *   <http://cpantest.grango.org/>

        The new CPAN Testers Wiki (thanks Barbie!)

    *   <http://lists.cpan.org/showlist.cgi?name=cpan-testers>

        The cpan-testers mailing list

    Test::Reporter itself--as a project--also has several links for your
    visiting enjoyment:

    *   <http://code.google.com/p/test-reporter/>

        Test::Reporter's master project page

    *   <http://groups.google.com/group/test-reporter>

        Discussion group for Test::Reporter

    *   <http://code.google.com/p/test-reporter/w/list>

        The Wiki for Test::Reporter

    *   <http://eclipse.resort.org/git/gitweb.cgi?p=test-reporter.git>

        Test::Reporter's public git source code repository.

    *   <http://search.cpan.org/dist/Test-Reporter/>

        Test::Reporter on CPAN

    *   <http://code.google.com/p/test-reporter/issues/list>

        UNFORTUNATELY, WE ARE UNABLE TO ACCEPT TICKETS FILED WITH RT.

        Please file all bug reports and enhancement requests at our Google
        Code issue tracker. Thank you for your support and understanding.

    *   <http://backpan.cpan.org/authors/id/F/FO/FOX/>

    *   <http://backpan.cpan.org/authors/id/A/AF/AFOXSON/>

        If you happen to--for some strange reason--be looking for primordial
        versions of Test::Reporter, you can almost certainly find them at
        the above 2 links.

METHODS
    *   address

        Optional. Gets or sets the e-mail address that the reports will be
        sent to. By default, this is set to cpan-testers@perl.org. You
        shouldn't need this unless the CPAN Tester's change the e-mail
        address to send report's to.

    *   comments

        Optional. Gets or sets the comments on the test report. This is most
        commonly used for distributions that did not pass a 'make test'.

    *   debug

        Optional. Gets or sets the value that will turn debugging on or off.
        Debug messages are sent to STDERR. 1 for on, 0 for off. Debugging
        generates very verbose output and is useful mainly for finding bugs
        in Test::Reporter itself.

    *   dir

        Optional. Defaults to the current working directory. This method
        specifies the directory that write() writes test report files to.

    *   distribution

        Gets or sets the name of the distribution you're working on, for
        example Foo-Bar-0.01. There are no restrictions on what can be put
        here.

    *   edit_comments

        Optional. Allows one to interactively edit the comments within a
        text editor. comments() doesn't have to be first specified, but it
        will work properly if it was. Accepts an optional hash of arguments:

        *   suffix

            Optional. Allows one to specify the suffix ("extension") of the
            temp file used by edit_comments. Defaults to '.txt'.

    *   errstr

        Returns an error message describing why something failed. You must
        check errstr() on a send() in order to be guaranteed delivery. This
        is optional if you don't intend to use Test::Reporter to send
        reports via e-mail, see 'send' below for more information.

    *   from

        Optional. Gets or sets the e-mail address of the individual
        submitting the test report, i.e. "afoxson@pobox.com (Adam Foxson)".
        This is mostly of use to testers running under Windows, since
        Test::Reporter will usually figure this out automatically.
        Alternatively, you can use the MAILADDRESS environmental variable to
        accomplish the same.

    *   grade

        Gets or sets the success or failure of the distributions's 'make
        test' result. This must be one of:

          grade     meaning
          -----     -------
          pass      all tests passed
          fail      one or more tests failed
          na        distribution will not work on this platform
          unknown   tests did not exist or could not be run

    *   mail_send_args -- DEPRECATED

        Kept for backwards compatibility. Use "transport_args" instead.

        Optional. If you have MailTools installed and you want to have it
        behave in a non-default manner, parameters that you give this method
        will be passed directly to the constructor of Mail::Mailer. See
        Mail::Mailer and Mail::Send for details.

    *   message_id

        Returns an automatically generated Message ID. This Message ID will
        later be included as an outgoing mail header in the test report
        e-mail. This was included to conform to local mail policies at
        perl.org. This method courtesy of Email::MessageID.

    *   mx

        Optional. Gets or sets the mail exchangers that will be used to send
        the test reports. If you override the default values make sure you
        pass in a reference to an array. By default, this contains the MX's
        known at the time of release for perl.org. If you do not have
        Mail::Send installed (thus using the Net::SMTP interface) and do
        have Net::DNS installed it will dynamically retrieve the latest
        MX's. You really shouldn't need to use this unless the hardcoded
        MX's have become wrong and you don't have Net::DNS installed.

    *   new

        This constructor returns a Test::Reporter object. It will optionally
        accept named parameters for: mx, address, grade, distribution, from,
        comments, via, timeout, debug, dir, perl_version, transport and
        transport_args.

    *   perl_version

        Returns a hashref containing _archname, _osvers, and _myconfig based
        upon the perl that you are using. Alternatively, you may supply a
        different perl (path to the binary) as an argument, in which case
        the supplied perl will be used as the basis of the above data.

    *   report

        Returns the actual content of a report, i.e. "This distribution has
        been tested as part of the cpan-testers...". 'comments' must first
        be specified before calling this method, if you have comments to
        make and expect them to be included in the report.

    *   send

        Sends the test report to cpan-testers@perl.org. You must check
        errstr() on a send() in order to be guaranteed delivery.
        Technically, send() is optional, as you may use Test::Reporter to
        only obtain the 'subject' and 'report' without sending an e-mail at
        all, although that would be unusual.

    *   subject

        Returns the subject line of a report, i.e. "PASS Mail-Freshmeat-1.20
        Darwin 6.0". 'grade' and 'distribution' must first be specified
        before calling this method.

    *   timeout

        Optional. Gets or sets the timeout value for the submission of test
        reports. Default is 120 seconds.

    *   transport

        Optional. Gets or sets the transport type. The transport type
        argument is refers to a 'Test::Reporter::Transport' subclass. The
        default is 'Net::SMTP', which uses the
        [Test::Reporter::Transport::Net::SMTP] class.

        You can add additional arguments after the transport selection.
        These will be passed to the constructor of the lower-level
        transport. This can be used to great effect for all manner of fun
        and enjoyment. ;-) See "transport_args".

        If Net::SMTP::TLS is used, 'Username' and 'Password' key-value
        transport arguments must be provided.

         $reporter->transport( 
             'Net::SMTP::TLS', Username => 'jdoe', Password => '123' 
         );

        If the 'HTTP' transport is used, two additional arguments are
        required: a URL to a Test::Reporter::HTTPGateway compatible server
        and an (optional) API key.

         $reporter->transport( 
             'HTTP', 'http://example.com/reporter-gateway/', '123456' 
         );

        This is not designed to be an extensible platform upon which to
        build transport plugins. That functionality is planned for the
        next-generation release of Test::Reporter, which will reside in the
        CPAN::Testers namespace.

    *   transport_args

        Optional. Gets or sets transport arguments that will used in the
        constructor for the selected transport, as appropriate.

    *   via

        Optional. Gets or sets the value that will be appended to
        X-Reported-Via, generally this is useful for distributions that use
        Test::Reporter to report test results. This would be something like
        "CPANPLUS 0.036".

    *   write and read

        These methods are used in situations where you test on a machine
        that has port 25 blocked and there is no local MTA. You use write()
        on the machine that you are testing from, transfer the written test
        reports from the testing machine to the sending machine, and use
        read() on the machine that you actually want to submit the reports
        from. write() will write a file in an internal format that contains
        'From', 'Subject', and the content of the report. The filename will
        be represented as:
        grade.distribution.archname.osvers.seconds_since_epoch.pid.rpt.
        write() uses the value of dir() if it was specified, else the cwd.

        On the machine you are testing from:

          my $reporter = Test::Reporter->new
          (
            grade => 'pass',
            distribution => 'Test-Reporter-1.16',
          )->write();

        On the machine you are submitting from:

          my $reporter;
          $reporter = Test::Reporter->new()->read('pass.Test-Reporter-1.16.i686-linux.2.2.16.1046685296.14961.rpt')->send() || die $reporter->errstr(); # wrap in an opendir if you've a lot to submit

        write() also accepts an optional filehandle argument:

          my $fh; open $fh, '>-';  # create a STDOUT filehandle object
          $reporter->write($fh);   # prints the report to STDOUT

CAVEATS
    If you experience a long delay sending mail with Test::Reporter, you may
    be experiencing a wait as Test::Reporter attempts to determine your
    email domain. Setting the MAILDOMAIN environment variable will avoid
    this delay.

COPYRIGHT
     Copyright (C) 2008 David A. Golden.
     Copyright (C) 2002, 2003, 2004, 2005, 2006, 2007, 2008 Adam J. Foxson.
     Copyright (C) 2004, 2005 Richard Soderberg.
     All rights reserved.

LICENSE
    This program is free software; you may redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    *   perl

    *   Config

    *   Net::SMTP

    *   Net::SMTP::TLS

    *   File::Spec

    *   File::Temp

    *   Net::Domain

        This is optional. If it's installed Test::Reporter will try even
        harder at guessing your mail domain.

    *   Net::DNS

        This is optional. If it's installed Test::Reporter will dynamically
        retrieve the mail exchangers for perl.org, instead of relying on the
        MX's known at the time of this release.

    *   Test::Reporter::HTTPGateway

        This is optional. It provides a web API for the 'HTTP' transport
        method.

AUTHORS
     Adam J. Foxson <afoxson@pobox.com>
     David Golden <dagolden@cpan.org>
     Kirrily "Skud" Robert <skud@cpan.org>
     Ricardo Signes <rjbs@cpan.org>
     Richard Soderberg <rsod@cpan.org>
     Kurt Starsinic <Kurt.Starsinic@isinet.com>



Changes
Revision history for Perl extension Test::Reporter.

For the complete changelog, visit:
http://github.com/dagolden/test-reporter/commits/master

{{$NEXT}}

1.62      2015-04-06 12:02:21-04:00 America/New_York

        - no changes from 1.61-TRIAL

1.61      2015-03-30 10:13:06-04:00 America/New_York (TRIAL RELEASE)

        [FIXED}

        - protect whitespace in perl path from shell

1.60      2013-11-04 13:49:54 America/New_York

        [FIXED]

        - read() now fixes up the architecture name

        - fixed documentation example for transport arguments
          involving File transport

1.59      2013-06-05 23:10:30 America/New_York


        [FIXED]

        - Fixed bug in finding right perl for -V output
          [rt.cpan.org #85823]

1.58      2011-07-19 12:07:53 America/New_York

        - Deprecates email-transport related attributes and methods

        - Adds new 'Null' transport as default

        - Removes legacy transport classes
          (See Test::Reporter::Transport::Legacy if needed)

        - Removes 'cpantest' command line program

1.57      2010-03-26 00:13:18 EST5EDT

        - Includes full 'perl -V' output, not just Config::mychanges

        - Adds new 'X-Test-Reporter-Perl' header with Perl version 
          and optional "RCx" suffix to positively identify tests from
          RC Perls

        - Minimum perl version is now 5.006, matching the rest of the Perl toolchain

1.56      2010-03-14 23:26:00 EST5EDT

        - Convert line endings during read() [Issue identified
          by Serguei Trouchelle]

1.55      2010-03-12 11:44:05 EST5EDT

        - Preserve distfile through write() and read() 
          [patch by Dan Collins]

        - Regenerate _myconfig data during read() method

        - Switched to Dist::Zilla for release management

1.54 Sun May 31 21:39:12 EDT 2009

        - POD updates.

        - Update Module::Install to 0.91.

1.53_03 Thu Apr 16 12:30:49 EDT 2009

        - Transmit report body line-by-line with Net::SMTP (or Net::SMTP::TLS).

1.53_02 Wed Apr 15 13:30:53 EDT 2009

        - With Net::SMTP transport, if lines are longer than 100 characters,
          the email will be encoded with quoted-printable.

1.53_01 Tue Apr  7 11:01:37 EDT 2009

        - Added optional 'distfile' argument to support
          T::R::Transport::Metabase.

        - Reports link to the wiki rather than www.cpantesters.org; also added
          cpan-testers-discuss as place to send Q's.

        - Improved diagnostic output if editor fails to launch.

        - Changed Makefile.Pl to Module::Install.

1.5203 Tue Jan 20 15:25:07 EST 2009

        - Updates CPAN Testers urls to www.cpantesters.org (David Westbrook).

1.5202 Mon Nov  3 11:48:44 EST 2008

        - Fixed bug in Mail::Send transport (GoogleCode issue 19; patch by
          Moshe Kamensky).

1.5201 Tue Sep 23 11:18:34 EDT 2008

        - All tests now use Test::More (not Test) for better diagnostics.

        - Removed unused signature test.

1.52 Mon Sep  8 11:45:12 EDT 2008

        - Fixed extraneous-newlines bug in reading reports saved to a file
          (Sleven Rezick).

1.51_02 Sun Sep  7 23:24:47 EDT 2008

        - Cache mail domain; shorten mail domain timeout; add 'smtp' as a
          hostname to check (KANE).

1.51_01 Sat Sep  6 14:34:13 EDT 2008

        - Avoid exclusive lock on tempfile for editing report (Slaven Rezick).

1.50 Sat Sep  6 14:04:22 EDT 2008

        - Deprecated sending reports to anywhere other than the cpan-testers
          mailing list.  This means that authors will no longer be copied on
          report emails by individual testers. (CPAN Testers is switching to
          an opt-in philosophy with central author notifications).

        - Doc fix on usage of read() in Test::Reporter::Transport::File.

1.4002 Wed Jun 25 17:21:10 EDT 2008

        - Fixed bug in error handling when using TLS transport.

1.4001 Wed May 21 12:34:28 EDT 2008

        - Fixed bug where send() would crash when no explicit transport was
          set.

1.40 Sun May 18 14:54:46 EDT 2008

        - No change from 1.39_06.

1.39_06 Thu Apr 17 06:15:45 EDT 2008

        - Fixed bug setting transport_args in constructor.

1.39_05 Wed Apr  9 12:43:55 EDT 2008

        - Removed use of File::Temp->newdir() for backwards compatibility with
          older versions of File::Temp.

1.39_04 Mon Apr  7 17:25:35 EDT 2008

        - Added Test::Reporter::Transport::File subclass.

1.39_03 Mon Apr  7 10:00:00 EDT 2008

        - Added documentation for transport.

1.39_02 Sun Apr  6 14:40:00 EDT 2008

        - Refactored transport into Test::Reporter::Transport subclasses.

1.39_01 Sun Apr  6 01:16:10 EDT 2008

        - Whoops! Devel versions are supposed to be odd.

        - t/2-http-transport.t left out of MANIFEST.

        - Copyright notice updates.

1.38_01 Sat Apr  5 18:00:00 EDT 2008

        - Add support for 'HTTP' transport to a Test::Reporter::HTTPGateway
          server.

1.38 Tue Oct 09 21:00:00 EDT 2007

        - Additional compatibility improvements for VMS. Many thanks to Peter
          (Stig) Edwards!

1.36 Mon Oct 08 21:00:00 EDT 2007

        - Compat for perl-5.005_05 per David Golden and Slaven Rezic.

        - Per 29690@RT cleaning up cpantest's -dump output (thanks janus!).

        - Per 9@Google compat fixes for VMS courtesy of Peter (Stig) Edwards.

1.34 Sun Aug 12 23:00:00 EDT 2007

        - It is now possible to override the automatic selection of a transport
          by making use of the transport() method (per WYANT and ANDK).

        - A dozen new tests added to support transport().

        - Internal debugging improvements.

        - Various documentation improvements.

        - Removing subversion artifacts; This project is now under git control.

1.32 Sun Jul 29 06:30:00 EDT 2007

        - Fixed an issue (8@Google) whereby test reports had Date headers that
          were not RFC-822 complaint (reported by Serguei Trouchelle).


lib/Test/Reporter/Mail/Util.pm
# Test::Reporter::Mail::Util.pm
#
# Copyright (c) 1995-2001 Graham Barr <gbarr@pobox.com>. All rights reserved.
# Copyright (c) 2002-2003 Mark Overmeer <mailtools@overmeer.net>
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.

package Test::Reporter::Mail::Util;

use strict;
use vars qw($VERSION @ISA @EXPORT $mailaddress $maildomain $realname);
use Exporter;

$VERSION = '1.58';
@ISA = qw(Exporter);
@EXPORT = qw(_realname _maildomain _mailaddress $mailaddress $maildomain $realname);

local $^W;

sub _realname {
	my $self = shift;

	warn __PACKAGE__, ": _realname\n" if $self->debug();

	return $realname if defined $realname;

	$realname = '';
	$realname                                    =
		eval {(split /,/, (getpwuid($>))[6])[0]} ||
		$ENV{QMAILNAME}                          ||
		$ENV{REALNAME}                           ||
		$ENV{USER};

	return $realname;
}

sub _maildomain {
	my $self = shift;

	warn __PACKAGE__, ": _maildomain\n" if $self->debug();

	return $maildomain if defined $maildomain;

	$maildomain = $ENV{MAILDOMAIN};

	return $maildomain if defined $maildomain;

	local *CF;
	local $_;
	my @sendmailcf = qw(
		/etc
		/etc/sendmail
		/etc/ucblib
		/etc/mail
		/usr/lib
		/var/adm/sendmail
	);
	my $config = (grep(-r, map("$_/sendmail.cf", @sendmailcf)))[0];

	if (defined $config && open(CF, $config)) {
		my %var;
		while (<CF>) {
			if (my ($v, $arg) = /^D([a-zA-Z])([\w.\$\-]+)/) {
				$arg =~ s/\$([a-zA-Z])/exists $var{$1} ? $var{$1} : '$'.$1/eg;
				$var{$v} = $arg;
			}
		}

		close(CF) || die $!;
		$maildomain = $var{j} if defined $var{j};
		$maildomain = $var{M} if defined $var{M};

		$maildomain = $1
			if($maildomain && $maildomain =~ m/([A-Za-z0-9](?:[\.\-A-Za-z0-9]+))/ );

		return $maildomain if defined $maildomain;
	}

	if (open(CF,"/usr/lib/smail/config")) {
		while (<CF>) {
			if (/\A\s*hostnames?\s*=\s*(\S+)/) {
				$maildomain = (split(/:/,$1))[0];
				last;
			}
		}
		close(CF) || die $!;

		return $maildomain if defined $maildomain;
	}

	my $host;

	for $host (qw(mailhost localhost)) {
		my $smtp = Net::SMTP->new($host);

		if (defined $smtp) {
			$maildomain = $smtp->domain;
			$smtp->quit;
			last;
		}
	}

	unless (defined $maildomain) {
		if ($self->_have_net_domain()) {
			$maildomain = Net::Domain::domainname();
		}
	}

	$maildomain = "localhost" unless defined $maildomain;

	return $maildomain;
}

sub _mailaddress {
	my $self = shift;

	warn __PACKAGE__, ": _mailaddress\n" if $self->debug();

	return $self->{_from} if defined $self->{_from};
	return $mailaddress if defined $mailaddress;

	my $realname = $self->_realname();

	$mailaddress = $ENV{MAILADDRESS};

	unless ($mailaddress || $^O ne 'MacOS') {
		my %InternetConfig;
		require Mac::InternetConfig;
		Mac::InternetConfig->import();
		$mailaddress = $InternetConfig{kICEmail()};
	}

	$mailaddress ||=
		$ENV{USER}          ||
		$ENV{LOGNAME}       ||
		eval {getpwuid($>)} ||
		"postmaster";

	$mailaddress .= '@' . $self->_maildomain() unless $mailaddress =~ /\@/;
	$mailaddress =~ s/(^.*<|>.*$)//g;

	if ($realname) {
		$mailaddress = "$mailaddress ($realname)";
	}

	return $mailaddress;
}

1;

__DATA__
=pod

=head1 NAME

Test::Reporter::Mail::Util - mail utility functions

=head1 SYNOPSIS

use Test::Reporter::Mail::Util qw( ... );

=head1 DESCRIPTION

This package provides several mail related utility functions. Any function
required must by explicitly listed on the use line to be exported into
the calling package.

=head2 maildomain()

Attempt to determine the current uers mail domain string via the following
methods

=over 4

=item *  Look for the MAILDOMAIN enviroment variable, which can be set from outside the program.

=item *  Look for a sendmail.cf file and extract DH parameter

=item *  Look for a smail config file and usr the first host defined in hostname(s)

=item *  Try an SMTP connect (if Net::SMTP exists) first to mailhost then localhost

=item *  Use value from Net::Domain::domainname (if Net::Domain exists)

=back

=head2 mailaddress()

Return a guess at the current users mail address. The user can force
the return value by setting the MAILADDRESS environment variable.

=head2 realname()

Attempts to get the reporter's real name, i.e. whatever happens to be in
the GCOS field.

=head1 AUTHOR

Graham Barr.

Maintained by Mark Overmeer <mailtools@overmeer.net>

=head1 COPYRIGHT

Copyright (c) 2002-2003 Mark Overmeer, 1995-2001 Graham Barr. All rights
reserved. This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Reporter.pm
# $Revision: 1.13 $
# $Id: Reporter.pm,v 1.13 2002/08/12 07:18:48 afoxson Exp $

# Test::Reporter - reports test results to the CPAN testing service
# Copyright (c) 2002 Adam J. Foxson. All rights reserved.

# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

package Test::Reporter;

use strict;
use Config;
use Carp;
use Net::SMTP;
use vars qw($VERSION $AUTOLOAD);

($VERSION) = '$Revision: 1.13 $' =~ /\s+(\d+\.\d+)\s+/;

local $^W;

sub new {
	my $type  = shift;
	my $class = ref($type) || $type;
	my $self  = {
		'_mx'           => ['perlmail.valueclick.com', 'lux.valueclick.com'],
		'_address'      => 'cpan-testers@perl.org',
		'_grade'        => undef,
		'_distribution' => undef,
		'_report'       => undef,
		'_subject'      => undef,
		'_from'         => undef,
		'_comments'     => '',
		'_errstr'       => '',
		'_timeout'      => 120,
	};

	bless $self, $class;

	$self->{_attr} = {   
		map {$_ => 1} qw(   
			_distribution _comments _errstr _timeout
		)
	};

	return $self;
}

sub subject {
	my $self = shift;

	croak __PACKAGE__, ": subject: grade and distribution must first be set\n"
		if not defined $self->{_grade} or not defined $self->{_distribution};

	my $subject = uc($self->{_grade}) . ' ' . $self->{_distribution} .
		" $Config{archname} $Config{osvers}";

	return $self->{_subject} = $subject;
}

sub report {
	my $self   = shift;
	my $report = qq(
		This distribution has been tested as part of the cpan-testers
		effort to test as many new uploads to CPAN as possible.  See
		http://testers.cpan.org/

		Please cc any replies to cpan-testers\@perl.org to keep other
		test volunteers informed and to prevent any duplicate effort.
	);

	$report =~ s/\n//;
	$report =~ s/\t{2}//g;

	if (not $self->{_comments}) {
		$report .= "\n\n--\n\n";
	}
	else {
		$report .= "\n--\n" . $self->{_comments} . "\n--\n\n";
	}

	$report .= Config::myconfig();

	chomp $report;
	chomp $report;

	return $self->{_report} = $report;
}

sub grade {
	my ($self, $grade) = @_;
	my %grades = (
		'pass'    => "all tests passed",
		'fail'    => "one or more tests failed",
		'na'      => "distribution will not work on this platform",
		'unknown' => "distribution did not include tests",
	);

	return $self->{_grade} if scalar @_ == 1;

	croak __PACKAGE__, ":grade: '$grade' is invalid, choose from: " .
		join ' ', keys %grades unless $grades{$grade};

	return $self->{_grade} = $grade;
}

sub send {
	my ($self, @recipients) = @_;

	$self->from();
	$self->subject();
	$self->report();

	return unless $self->_verify();
	return $self->_send_smtp(@recipients);
}

sub _verify {
	my $self = shift;
	my @undefined;

	for my $key (keys %{$self}) {
		push @undefined, $key if not defined $self->{$key};
	}

	$self->errstr(__PACKAGE__ . ": Missing values for: " .
		join ', ', map {$_ =~ /^_(.+)$/} @undefined) if
		scalar @undefined > 0;
	$self->errstr() ? return 0 : return 1;
}

sub _send_smtp {
	my $self       = shift;
	my $helo       = $self->_maildomain();
	my $from       = $self->from();
	my @recipients = @_;
	my $success    = 0;
	my $smtp       = Net::SMTP->new((@{$self->{_mx}})[0], Hello => $helo,
						Timeout => $self->{_timeout});
	my $recipients;

	if (@recipients) {
		$recipients = join ', ', @recipients;
		chomp $recipients;
		chomp $recipients;
	}

	$success += $smtp->mail($from);
	$success += $smtp->to($self->{_address});
	$success += $smtp->cc(@recipients) if @recipients;
	$success += $smtp->data();
	$success += $smtp->datasend("From: $from\n");
	$success += $smtp->datasend("To: ", $self->{_address}, "\n");
	$success += $smtp->datasend("Cc: $recipients\n") if @recipients;
	$success += $smtp->datasend("Subject: ", $self->subject(), "\n");
	$success +=
		$smtp->datasend("X-reported-via: Test::Reporter $VERSION\n");
	$success += $smtp->datasend("\n");
	$success += $smtp->datasend($self->report());
	$success += $smtp->dataend();
	$success += $smtp->quit;

	if (@recipients) {
		$self->errstr(__PACKAGE__ .
			': Unable to send test report to one or more recipients') if
				$success != 13;
	}
	else {
		$self->errstr(__PACKAGE__ . ': Unable to send test report') if
			$success != 11;
	}

	$self->errstr() ? 0 : 1;
}

sub _realname {
	my $self     = shift;
	my $realname = '';

	$realname                                    =
		eval {(split /,/, (getpwuid($>))[6])[0]} ||
		$ENV{QMAILNAME}                          ||
		$ENV{REALNAME}                           ||
		$ENV{USER};

	return $realname;
}

# Adapted from Mail::Util
sub _maildomain {
	my $self       = shift;
	my @sendmailcf = qw(
		/etc
		/etc/sendmail
		/etc/ucblib
		/etc/mail
		/usr/lib
		/var/adm/sendmail
	);
	my $config = (grep(-r, map("$_/sendmail.cf", @sendmailcf)))[0];
	my $domain;

	if (defined $config && open(CF, $config)) {
		my %var;
		while (<CF>) {
			if (my ($v, $arg) = /^D([a-zA-Z])([\w.\$\-]+)/) {
				$arg =~ s/\$([a-zA-Z])/exists $var{$1} ? $var{$1} : '$'.$1/eg;
				$var{$v} = $arg;
			}
		}

		close(CF) || die $!;
		$domain = $var{j} if defined $var{j};
		$domain = $var{M} if defined $var{M};
		$domain = $var{S} if defined $var{S};
		return $domain if defined $domain;
	}

	if (open(CF,"/usr/lib/smail/config")) {
		while (<CF>) {
			if (/\A\s*hostnames?\s*=\s*(\S+)/) {
				$domain = (split(/:/,$1))[0];
				last;
			}
		}
		close(CF);

		return $domain if defined $domain;
	}

	my $host;

	for $host (qw(mailhost localhost)) {
		my $smtp = eval {Net::SMTP->new($host)};

		if (defined $smtp) {
			$domain = $smtp->domain;
			$smtp->quit;
			last;
		}
	}

	unless (defined $domain) {
		if (eval {require Net::Domain}) {
			$domain = Net::Domain::domainname();
		}
	}

	$domain = "localhost" unless defined $domain;

	return $domain;
}

# Adapted from Mail::Util
sub _mailaddress {
	my $self     = shift;
	my $realname = $self->_realname();
	my $mailaddress;

	return $self->{_from} if defined $self->{_from};

	$mailaddress            =
		$ENV{MAILADDRESS}   ||
		$ENV{USER}          ||
		$ENV{LOGNAME}       ||
		eval {getpwuid($>)} ||
		"postmaster";

	$mailaddress .= '@' . $self->_maildomain() unless $mailaddress =~ /\@/;
	$mailaddress =~ s/(^.*<|>.*$)//g;

	if ($realname) {
		$mailaddress = "$mailaddress ($realname)";
	}
	else {
		$mailaddress = $mailaddress;
	}

	return $mailaddress;
}

sub from {
	my $self = shift;

	if (@_) {
		$self->{_from} = shift;
	}
	else {
		$self->{_from} = $self->_mailaddress();
	}

	return $self->{_from};
}

sub AUTOLOAD {
	my $self               = $_[0];
	my ($package, $method) = ($AUTOLOAD =~ /(.*)::(.*)/);

	return if $method =~ /^DESTROY$/;

	unless ($self->{_attr}->{"_$method"}) {
		croak __PACKAGE__ . ": No such method: $method; aborting";
	}

	my $code = q{
		sub {   
			my $self = shift;
			$self->{_METHOD} = shift if @_;
			return $self->{_METHOD};
		}
	};

	$code =~ s/METHOD/$method/g;

	{
		no strict 'refs';
		*$AUTOLOAD = eval $code;
	}

	goto &$AUTOLOAD;
}

1;

__DATA__
=pod

=head1 NAME

Test::Reporter - reports test results to the CPAN testing service

=head1 SYNOPSIS

  use Test::Reporter;

  my $reporter = Test::Reporter->new();

  $reporter->grade('pass');
  $reporter->distribution('Mail-Freshmeat-1.20');
  $reporter->send() || die $reporter->errstr();

  # or

  my $reporter = Test::Reporter->new();

  $reporter->grade('fail');
  $reporter->distribution('Mail-Freshmeat-1.20');
  $reporter->comments('output of a failed make test goes here...');
  $reporter->send('afoxson@pobox.com') || die $reporter->errstr();

  NOTE: THIS VERSION OF Test::Reporter SHOULD BE CONSIDERED BETA.
  THE INTERFACE MAY CHANGE.

=head1 DESCRIPTION

Test::Reporter reports the test results of any given distribution to the
CPAN testing service. See B<http://testers.cpan.org/> for details.

=head1 METHODS

=over 4

=item * B<new>

This constructor returns a Test::Reporter object. It currently accepts no
parameters, but will likely take named parameters in a future version.

=item * B<subject>

Returns the subject line of a report, i.e.
"PASS Mail-Freshmeat-1.20 Darwin 6.0". 'grade' and 'distribution' must
first be specified before calling this method.

=item * B<report>

Returns the actual content of a report, i.e.
"This distribution has been tested as part of the cpan-testers...". 
'comments' must first be specified before calling this method, if you have
comments to make and expect them to be included in the report.

=item * B<comments>

Gets or sets the comments on the test report. This is optional, and most
commonly used for distributions that did not pass a 'make test'.

=item * B<errstr>

Returns an error message describing why something failed. You must check
errstr() on a send() in order to be guaranteed delivery.

=item * B<from>

Gets or sets the e-mail address of the individual submitting the test
report, i.e. "Adam J. Foxson <afoxson@pobox.com>". This is mostly
of use to testers running under Windows, since Test::Reporter will
usually figure this out automatically.

=item * B<grade>

Gets or sets the success or failure of the distributions's 'make test'
result. This must be one of:

  grade     meaning
  -----     -------
  pass      all tests passed
  fail      one or more tests failed
  na        distribution will not work on this platform
  unknown   distribution did not include tests

=item * B<distribution>

Gets or sets the name of the distribution you're working on, for example
Foo-Bar-0.01. There are no restrictions on what can be put here.

=item * B<send>

Sends the test report to cpan-testers@perl.org and cc's the e-mail to the
specified recipients, if any. If you do specify recipients to be cc'd
be sure that you use the author's @cpan.org address otherwise they will
not be delivered.

=item * B<timeout>

Gets or sets the timeout value for the submission of test reports. 

=back

=head1 TODO

  - More detailed error messages and reporting.
  - More tests.
  - Possibly use Net::DNS to get MX's for perl.org.
  - Cycle through all available MX servers, instead of just the first.
  - Encoding of data may be necessary.
  - Allow the constructor to take named parameters.

=head1 COPYRIGHT

Copyright (c) 2002 Adam J. Foxson. All rights reserved.

=head1 LICENSE

This program is free software; you may redistribute it
and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

Adam J. Foxson <afoxson@pobox.com>, with much deserved credit to
Kirrily "Skud" Robert <skud@cpan.org>, and
Kurt Starsinic E<lt>F<Kurt.Starsinic@isinet.com>E<gt> for predecessor
versions (CPAN::Test::Reporter, and cpantest respectively).

=cut
